src/lib/shim/shim_syscall.c

```diff
@@ -263,8 +267,20 @@ static SysCallReg _shim_emulated_syscall_event(const ShimEvent* syscall_event) {
                 // (e.g. `kill(getpid(), signo)`), or may have been sent by another thread
                 // while this one was blocked in a syscall.
                 ShimShmemHostLock* host_lock = shimshmemhost_lock(shim_hostSharedMem());
-                _shim_process_signals(host_lock);
+                const bool sigactionsHadSaRestart = _shim_process_signals(host_lock);
                 shimshmemhost_unlock(shim_hostSharedMem(), &host_lock);
+                if (rv.as_i64 == -EINTR) {
+                    // Syscall was interrupted by a signal. Consider restarting. See signal(7).
+                    const bool syscallSupportsSaRestart =
+                        res.event_data.syscall_complete.restartable;
```

Comment:
I think there should be a comment somewhere around here explaining that `syscallhandler_make_syscall()` checks for pending signals ahead of time to convert blocking syscall results to `EINTR` syscall results. It's a little confusing how the syscall result changes throughout shadow. For example, the syscall handler may decide to block, and then shadow may decide to change that block into a `EINTR`, then the shim may decide to change that `EINTR` into a new syscall event for the same syscall (restarts the syscall).